### PR TITLE
feat: Add support for time-zone, 3 & 5 digit years: Cast from string to timestamp. 

### LINKED ISSUE #376: Cast string to timestamp remaining work
### What is the problem the feature request solves?

https://github.com/apache/datafusion-comet/pull/335 added initial support for casting string to timestamp, but the feature is still disabled by default. There are some follow on items to implement before we can enable by default:

- Add support for 3 and 5 digit years
- Testing with try_cast and ANSI
- Add timezone support (see advice in https://github.com/apache/datafusion-comet/pull/335#issuecomment-2091930125)
- Passing all fuzz tests
- Updating compatibility guide

### Describe the potential solution

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by akhilss99**
## Which issue does this PR close?

addresses timezone, 3 & 5 digit years' compatibility mentioned in #376 

## What changes are included in this PR?

Added 3 & 5 digit year, time-zone support.

## How are these changes tested?

Through Unit tests.

### CODE REVIEW

**Comment by andygrove** [comment]
Thanks @akhilss99. Could you run `make format` to fix the code formatting?

**Comment by akhilss99** [comment]
@andygrove, can you please run the CI build again? thanks!

**Comment by viirya** [comment]
Triggered.

File: docs/source/user-guide/compatibility.md [hunk 1]
```diff
@@ -116,6 +116,7 @@ The following cast operations are generally compatible with Spark except for the
 | string | long |  |
 | string | binary |  |
 | string | date | Only supports years between 262143 BC and 262142 AD |
+| string | timestamp |  |
```
Comment (andygrove) [inline docs/source/user-guide/compatibility.md:119]: To mark this cast as compatible, we would need to enable the `cast StringType to TimestampType` fuzz test in CometCastSuite, and this currently fails due to incompatibilities between Spark and Comet.

File: native/spark-expr/src/cast.rs [hunk 2]
```diff
@@ -1746,35 +1844,92 @@ mod tests {
     #[test]
     #[cfg_attr(miri, ignore)] // test takes too long with miri
     fn timestamp_parser_test() {
+        let tz = &timezone::Tz::from_str("UTC").unwrap();
         // write for all formats
         assert_eq!(
-            timestamp_parser("2020", EvalMode::Legacy).unwrap(),
+            timestamp_parser("2020", EvalMode::Legacy, tz).unwrap(),
```
Comment (andygrove) [inline native/spark-expr/src/cast.rs:1850]: Could you also add tests for 3-digit year

**Review by andygrove — COMMENTED** [review]
This looks like a nice improvement to the existing code, but doesn't fully close the issue.

I suspect that we may need to move away from the regex approach and port Spark's parsing code over (we did this for string to date already) to get full compatibility. This will likely be more performant as well.

I think this PR would still be worth merging though.

File: native/spark-expr/src/cast.rs (hunk 2 above)
Comment (akhilss99) [reply native/spark-expr/src/cast.rs:1850]: I've covered the test-cases for 3 digit years from L1877 - 1904, is there anything that I'm missing out?

File: docs/source/user-guide/compatibility.md (hunk 1 above)
Comment (akhilss99) [reply docs/source/user-guide/compatibility.md:119]: Got it, I mistook fuzz tests as running fuzz-test jar through SPARK-SUBMIT. Will revert the compatibility-guide changes.

**Comment by akhilss99** [comment]
Can anyone please run the build? 
thanks

**Comment by viirya** [comment]
Triggered again.

**Review by andygrove — APPROVED** [review]
Thanks @akhilss99

**Pull request merged**